### PR TITLE
BROOKLYN-406: report failure if coercion's fromMap throws exception

### PR DESCRIPTION
**Pull request by aledsage**

### CODE REVIEW

File: utils/common/src/main/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercerExtensible.java [hunk 1]
```diff
@@ -119,8 +120,23 @@ public static TypeCoercerExtensible newEmpty() {
         if (targetTypeToken.getType() instanceof ParameterizedType) {
             if (value instanceof Collection && Collection.class.isAssignableFrom(targetType)) {
                 result = tryCoerceCollection(value, targetTypeToken, targetType);
+                
+                if (result != null && result.isAbsent() && targetType.isInstance(value)) {
+                    log.warn("Failed to coerce collection from " + value.getClass().getName() + " to " + targetTypeToken  
+                            + "; returning uncoerced result to preserve (deprecated) backwards compatibility", 
+                            Maybe.getException(result));
+                    return Maybe.of((T)result);
```
Comment (neykov) [inline utils/common/src/main/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercerExtensible.java]: Should be `result.get()`

Comment (neykov) [inline utils/common/src/main/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercerExtensible.java]: Should be `return Maybe.of( (T) value );`, but I don't understand how it's different from letting the code execution continue. Same for change below.

File: utils/common/src/main/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercerExtensible.java [hunk 2]
```diff
@@ -209,12 +226,19 @@ public static TypeCoercerExtensible newEmpty() {
...
                     } catch (Exception e) {
-                        Maybe.absent(new ClassCoercionException("Cannot coerce type "+value.getClass()+" to "+targetType.getCanonicalName()+" ("+value+"): "+m.getName()+" adapting failed, "+e));
+                        exceptions.add(new ClassCoercionException("Cannot coerce type "+value.getClass()+" to "+targetType.getCanonicalName()+" ("+value+"): "+m.getName()+" adapting failed", e));
                     }
                 }
             }
         }
-        return null;
+        if (exceptions.isEmpty()) {
+            return null;
```
Comment (neykov) [inline utils/common/src/main/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercerExtensible.java]: Worth adding a comment somewhere on the notation (caller):
  * null - no errors, continue with fallbacks (i.e. not found)
  * absent - had some kind of exception, continue with fallbacks
  * present - coercion successful, return value

File: utils/common/src/test/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercionsTest.java [hunk 3]
```diff
@@ -344,6 +344,59 @@ public void testFrom() {
...
+        try {
+            coercer.tryCoerce("myval", WithFromThrowingException.class).get();
+            Asserts.shouldHaveFailedPreviously();
+        } catch (ClassCoercionException e) {
+            Asserts.expectedFailureContains(e, expectedGenericErr, expectedSpecificErr);
+        }
+    }
+
+    // TODO Asserting this undesirable behaviur, to preserve backwards compatibility!
+    // Would much prefer that we fail-fast. However, I worry that some entity's jave declared
```
Comment (neykov) [inline utils/common/src/test/java/org/apache/brooklyn/util/javalang/coerce/TypeCoercionsTest.java]: typo `jave`

**Comment by aledsage** [comment]
Thanks @neykov - incorporated those changes and rebased; merging now.

**Pull request merged**